### DISCUSSION TITLE
Custom theme from swizzled elements

### DISCUSSION THREAD

**Discussion by markusbkk** [facebook/docusaurus#8125] [Q&A]
I'm somewhat struggling with this.
I've been swizzling components for a while (including `Root`) and I'd now like to create my own theme from them.
Is there a documented way to do this?

EDIT: Figured it out. Will create an example project and link it here later.

**Comment by Advait251206** [comment]
Docusaurus doesn’t provide a formal “create your own theme package” workflow out of the box, but what you’re trying to do is absolutely possible using swizzling + re-exporting components in a structured way.

The key thing to understand is:

👉 Swizzling is meant for *overriding parts of an existing theme*, not for packaging a completely new theme  
…but you can still build a custom theme layer on top of it.

---

## Recommended approach (the “clean” way)

### 1. Treat swizzled components as your base layer

When you swizzle, you’re effectively copying:

```
@docusaurus/theme-classic
```

into:

```
src/theme/*
```

At this point, you already *own the implementation*, so you can start organizing it like a theme.

---

### 2. Create a “theme wrapper layer”

Instead of scattering overrides, group them:

```
src/theme/
  Layout/
  Navbar/
  Footer/
  Root/
  Blog/
  DocItem/
```

Each folder becomes your “custom theme implementation”.

---

### 3. Reuse original theme internally (important)

Inside swizzled components, you can still import original components:

```js
import OriginalNavbar from '@theme-original/Navbar';
```

This allows:

- extending instead of replacing
- gradual customization
- keeping upstream compatibility

---

### 4. Build your “theme API” pattern

If you want a real reusable theme, create an internal structure like:

```
src/theme-core/
  components/
  tokens/
  layout/
```

Then your swizzled components just consume that layer.

Example:

```js
import { useThemeTokens } from '@site/src/theme-core/tokens';
```

This is how you gradually turn overrides into a system.

---

## 5. If you want a true reusable theme (advanced)

Docusaurus supports **theming via npm packages**, but you must:

### Create a package like:
```
my-docusaurus-theme/
```

Structure:

```
src/
  theme/
  components/
  index.js
package.json
```

Then publish and use:

```bash
npm install my-docusaurus-theme
```

And in `docusaurus.config.js`:

```js
presets: [
  [
    'classic',
    {
      theme: {
        customCss: require.resolve('my-docusaurus-theme/src/index.css'),
      },
    },
  ],
],
```

But note:

👉 Full theme replacement still relies on swizzling behavior internally.

---

## Key limitation (important)

Docusaurus does NOT yet support:

- fully independent theme engines
- plugin-style theme injection system
- multiple themes switching at runtime

So everything is still built on top of:

> `@docusaurus/theme-classic` or custom overrides of it

---

## Practical recommendation

If your goal is:

### ✔ “Organized custom theme”
→ Use structured swizzled components + shared theme-core

### ✔ “Reusable theme across projects”
→ Move swizzled components into a separate npm package

### ❌ “Completely separate theme system”
→ Not currently supported in Docusaurus core

---

## Summary

What you’re trying to do is a valid path, but in Docusaurus terms:

> A “custom theme” is really a **well-structured set of swizzled overrides + shared abstractions**, not a standalone theme engine.

---

If this answer helped or pointed you in the right direction, I'd appreciate it if you could mark it as the accepted answer so it's easier for others with the same issue to find.

Also, if you found my contribution useful, I'd appreciate it if you could check out my GitHub profile, follow me, and star any repositories you find interesting.

GitHub: https://github.com/Advait251206
